presto-main/src/main/java/com/facebook/presto/server/security/KerberosAuthenticator.java

```diff
@@ -149,35 +128,88 @@ public Principal authenticate(HttpServletRequest request)
 
     private Optional<Principal> authenticate(String token)
     {
-        GSSContext context = doAs(loginContext.getSubject(), () -> gssManager.createContext(serverCredential));
-
         try {
-            byte[] inputToken = Base64.getDecoder().decode(token);
-            context.acceptSecContext(inputToken, 0, inputToken.length);
+            Session session = getSession();
+            LOG.debug("session remaining lift time: %d seconds", session.getRemainingLifetime());
+            GSSContext context = doAs(session.getLoginContext().getSubject(), () -> gssManager.createContext(session.getServerCredential()));
 
-            // We can't hold on to the GSS context because HTTP is stateless, so fail
-            // if it can't be set up in a single challenge-response cycle
-            if (context.isEstablished()) {
-                return Optional.of(new KerberosPrincipal(context.getSrcName().toString()));
-            }
-            LOG.debug("Failed to establish GSS context for token %s", token);
-        }
-        catch (GSSException e) {
-            // ignore and fail the authentication
-            LOG.debug(e, "Authentication failed for token %s", token);
-        }
-        finally {
             try {
-                context.dispose();
+                byte[] inputToken = Base64.getDecoder().decode(token);
+                context.acceptSecContext(inputToken, 0, inputToken.length);
+
+                // We can't hold on to the GSS context because HTTP is stateless, so fail
+                // if it can't be set up in a single challenge-response cycle
+                if (context.isEstablished()) {
+                    return Optional.of(new KerberosPrincipal(context.getSrcName().toString()));
+                }
+                LOG.debug("Failed to establish GSS context for token %s", token);
             }
             catch (GSSException e) {
-                // ignore
+                // ignore and fail the authentication
+                LOG.debug(e, "Authentication failed for token %s", token);
```

Comment:
would log.error can explain more in the log if the exception or errors happened?

Reply:
Yeah, maybe. Some tradeoff around this: GSSException is very general exception and we don't know if the log is a server error or user error. It would also print the token, which I believe shouldn't be present in the production log. So, I would keep it untouched as the open source code.